### PR TITLE
Fix a use-after-free detected by ASAN

### PR DESCRIPTION
**Pull request by zwaldowski**
I AGREE TO THE GITUP CONTRIBUTOR LICENSE AGREEMENT

### CODE REVIEW

**Comment by swisspol** [comment]
`GCGitPathFromFileSystemPath()` returns a pointer on the autorelease pool and the pool is not drained. Did you actually see a crash?

**Comment by zwaldowski** [comment]
Yes, this crashed with a debugger attached before I turned ASAN on. ASAN identifies it as a use after free of stack memory. By C rules, the storage for `&filePath` can freely be deallocated after the close of the `if (filePattern)` scope, not necessarily at the end of the function scope.

Added CLA.

**Comment by zwaldowski** [comment]
More concretely: [Objective-C Automatic Reference Counting (ARC) section 7.7: Interior pointers](https://clang.llvm.org/docs/AutomaticReferenceCounting.html#id52) only guarantees the originating object (here, the `NSString`) will survive until last use of the interior pointer. Combined with the above, Clang is technically allowed to release the string immediately upon assignment of `&filePath`.

File: GitUpKit/Core/GCDiff.m [hunk 1]
```diff
@@ -464,12 +468,11 @@ - (GCDiff*)_diffWithType:(GCDiffType)type
...
   }
-  if (filePattern) {
+  if (filePattern && filePath) {
     diffOptions.pathspec.count = 1;
-    const char* filePath = GCGitPathFromFileSystemPath(filePattern);
-    diffOptions.pathspec.strings = (char**)&filePath;
+    diffOptions.pathspec.strings = (char **)&filePath;
 
-    static NSCharacterSet* set = nil;
+    static NSCharacterSet *set = nil;
```
Comment (swisspol) [inline GitUpKit/Core/GCDiff.m]: Don't change.

File: GitUpKit/Core/GCDiff.m [hunk 2]
```diff
@@ -464,12 +468,11 @@ - (GCDiff*)_diffWithType:(GCDiffType)type
   if (options & kGCDiffOption_IgnoreAllSpaces) {
     diffOptions.flags |= GIT_DIFF_IGNORE_WHITESPACE;
   }
-  if (filePattern) {
+  if (filePattern && filePath) {
     diffOptions.pathspec.count = 1;
-    const char* filePath = GCGitPathFromFileSystemPath(filePattern);
-    diffOptions.pathspec.strings = (char**)&filePath;
+    diffOptions.pathspec.strings = (char **)&filePath;
```
Comment (swisspol) [inline GitUpKit/Core/GCDiff.m]: Don't change.

File: GitUpKit/Core/GCDiff.m [hunk 3]
```diff
@@ -464,12 +468,11 @@ - (GCDiff*)_diffWithType:(GCDiffType)type
   if (options & kGCDiffOption_IgnoreAllSpaces) {
     diffOptions.flags |= GIT_DIFF_IGNORE_WHITESPACE;
   }
-  if (filePattern) {
+  if (filePattern && filePath) {
```
Comment (swisspol) [inline GitUpKit/Core/GCDiff.m]: Change to `if (filePath) {`

File: GitUpKit/Core/GCDiff.m [hunk 4]
```diff
@@ -442,6 +442,10 @@ - (GCDiff*)_diffWithType:(GCDiffType)type
   GCDiff* gcDiff = nil;
   git_diff* diff = NULL;
 
+  // Make sure filePattern lives until the end of scope, otherwise Clang will
+  // deallocate the local used in `if (filePath)` below.
+  const char* filePath = GCGitPathFromFileSystemPath(filePattern);
```
Comment (swisspol) [inline GitUpKit/Core/GCDiff.m]: Change comment to `Make sure filePath lives until end of scope`

File: GitUpKit/Core/GCDiff.m (hunk 1 above)
Comment (zwaldowski) [inline GitUpKit/Core/GCDiff.m]: Apologies, these were done by Clang-Format re: [#1 in CONTRIBUTING](https://github.com/git-up/GitUp/blob/master/CONTRIBUTING.md#L9). See #326, I suppose.

**Comment by zwaldowski** [comment]
Rebased and changes made.

File: GitUpKit/Core/GCDiff.m [hunk 5]
```diff
@@ -442,6 +442,9 @@ - (GCDiff*)_diffWithType:(GCDiffType)type
   GCDiff* gcDiff = nil;
   git_diff* diff = NULL;
 
+  // Make sure filePattern lives until the end of scope
```
Comment (swisspol) [inline GitUpKit/Core/GCDiff.m]: Typo: should say "filePath"

**Pull request merged**